runtime/protocol/http-client-engines/http-client-engine-ktor/common/src/aws/smithy/kotlin/runtime/http/engine/ktor/DefaultHttpClientEngine.kt

```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+package aws.smithy.kotlin.runtime.http.engine.ktor
+
+import aws.smithy.kotlin.runtime.http.engine.*
+
+/**
+ * Specifies the common default http client of which platform specific [HttpClientEngine]'s actualize
+ *
+ * @param config Provides configuration for the DefaultHttpClientEngine
+ */
+expect class DefaultHttpClientEngine(config: HttpClientEngineConfig = HttpClientEngineConfig.Default) : HttpClientEngineBase
```

Comment:
Why rename this? `Default` tells me nothing about the underlying engine. It's based on ktor and even the artifact is still `http-client-engine-ktor`. The fact that it's chosen as the "default" in codegen shouldn't influence this name. In particular because when a customer chooses an explicit engine to override the default, those names should be explicit, e.g.

* `CrtHttpEngine`
* `NettyHttpEngine`
* `KtorHttpEngine`
* `CurlHttpEngine`
* etc
I'm not saying we will support all of those or anything just that as the number of possible engines we do support expands over time they ought to be named explicitly IMHO.

When we get to KMP support for all this I'll probably rework how we wire up the default anyway because I'd like it to be based on some form of #495 rather than needing to be explicitly set in codegen but I don't have enough to go on yet to make a recommendation w.r.t what that will look like eventually.

Perhaps we just `typealias DefaultHttpClientEngine = KtorEngine` if the name is important to everyone else.

Reply:
Interesting. What if we don't support some engine on some platform for some random reason (e.g., KTOR on JS). Won't having an `expect` then mean we can't compile for the platform that's missing engine implementations?

Reply:
It's not a requirement that any particular engine support all platforms or even more than one.

To be clear I'm not necessarily saying we shouldn't use `expect/actual` here in this case. KTOR has implementations for each of the platforms such that we could provide a KMP version of our HTTP engine abstraction using ktor. Which solves the issue Ken was trying to solve here (enabling us to generate a KMP client from smithy (the CRT engine is already KMP so we don't have this problem)).

Reply:
> Won't having an expect then mean we can't compile for the platform that's missing engine implementations?

`expect` applies to every `target` declared in the build. An artifact need not support _every_ possible KMP target. It will only of course be usable as a dependency for those specific targets.

I think what you're really getting at is if we eventually have an engine that only works on say `linuxX64` "won't that mean that a customer can't use that in common"? The answer is yes. However, if they are providing a custom HTTP engine that isn't the default then that is on them to manage. If they are using KMP they may need to write a little bit of `expect/actual` themselves to handle per/target configuration differences.

Our only requirement is that what we generate is KMP compatible. A specific HTTP engine implementation can't have the same requirement because obviously some engines are only available on some platforms. 

Happy to chat offline about this as well, it's a meaty subject.

Reply:
Regarding the name change, I renamed it because I didn't find `KtorEngine` to lend itself as well to forward-compat.  I choose `Default` as that seemed closest to the intention of the type...meaning the default HTTP client that the SDK ships with.  By using a generic name, we may have more flexibility in the future if we decide to use a different implementation as default.  In any case, since you seem to feel strongly about the existing name, I'll change it back.

If there are other `expect/actual` concerns in this thread please LMK, it isn't clear to me that there is.

Reply:
I think this is worth a discussion offline to align but no other concerns though.